WIP: Issue #290 - Increase overall data visibility (#36)

storrisi opened the pull request:
This would fix #290 , #187 

<img width="1157" alt="schermata 2018-10-03 alle 14 28 45" src="https://user-images.githubusercontent.com/7654979/46486526-3df6b880-c7fe-11e8-8dab-334d883d52b6.png">
<img width="1170" alt="schermata 2018-10-03 alle 16 35 42" src="https://user-images.githubusercontent.com/7654979/46486535-451dc680-c7fe-11e8-8f38-59dd1876fc28.png">
![ott-04-2018 17-38-38](https://user-images.githubusercontent.com/7654979/46486469-228bad80-c7fe-11e8-9382-9507395fce2f.gif)

NOTE:
- This has been developed using mocked data, so it needs the service implementation in order to be merged.
- In this PR some refactors of the FullDetailsView are coming from the implementation done in the PR #316

storrisi commented on src/components/FullDetailView/InnerDataSection.js:
```diff
@@ -0,0 +1,28 @@
+import PlaceholderRenderer from '../Renderers/PlaceholderRenderer'
+import MonacoEditorWrapper from '../MonacoEditorWrapper'
+
+const InnerDataSection = ({ value, name, type = 'json', actionButton = null }) => {
```
What i've done here is to move all the render functions that previously were into the FullDetailComponent.
Since the component was more than 380 lines of codes, i've moved into separate presentational components all the render functions, in order to have the code more clean.

All the components that are used specifically by FullDetailsView are placed in the same folder, other components that are reusable everywhere (like Labels, ButtonWithTooltip) are in a more generic folder

jeffmcaffer replied:
+1 on the refactoring. Great idea.

I disagree however with the grouping. These components just **happen** to currently only be used in the full details view. It is quite reasonable to use them elsewhere. Assuming there is no significant simplification or economy we get from positioning them as "part of" the full details view, what do we lose from positioning them more generally? IMO it is completely legit to imagine the HeaderSection or LicenseSection (for example) being used in some other context where we want to show details.

storrisi replied:
It makes sense, but in this case a different naming should be applied to those components.
`HeaderSection` could have a meaning if positioned under the FullDetailsView folder, but it hasn't the same meaning when used a reusable component.

storrisi commented on src/components/FullDetailView/HeaderSection.js:
```diff
@@ -0,0 +1,60 @@
+import Definition from '../../utils/definition'
+import ButtonWithTooltip from '../Renderers/ButtonWithTooltip'
+
+const HeaderSection = props => {
```
Actually this is only a presentational component, and given that it doesn't have any internal method or state, it doesn't need to implement a class Component

jeffmcaffer replied:
Need and should are different. All the other components are classes and have PropTypes specs etc.  There is something quite simplifying about just doing it all the same way. I get that this is a small/simple component but would also argue that some of the other components done in this style (e.g., `DescribedSection`) are less simple and that function should really be factored into smaller functions that are then brought together in the `render` method of a component class. 

This also helps people recognize duplicate code and opportunities for generalization.

This also helps with the documentation if we use PropTypes to spec the props. We can use another syntax/structure but then it is harder to cut and paste similar text around and we have two ways of doing something. 

"Consistent code is simpler code"

storrisi replied:
I totally agree with this, in most of the case i always prefer Class Components over functionals.
But in this specific case, i just wanted only to clean the FullDetailsView code (which was originally full of functional render components) and keep the focus on the issue.

Of course those components could be further improved, and i can definitely work on that.
I just want to make sure that we all want to address that in this PR.

storrisi commented:
I've updated the PR addressing your suggestions, and commented some methods in order to let this code mergable, even if the new API are not yet available.
Given that, it's not a WIP anymore.

jeffmcaffer commented on src/api/clearlyDefined.js:
```diff
@@ -60,10 +66,10 @@ export function getCuration(token, entity, params = {}) {
  * @param  {} token
  * @param  {} entity
- * @param  {} params
+ * @param  {object} params can contain property: pendingPrs, if true returns also curations not already merged
```
Not a fan of boolean arguments. better would be an "expand" enum that indicated what all to include. that is the way that the REST api would be written.

jeffmcaffer replied:
Yes, that is a correct interpretation for the REST api. The `getCuration` function should be similar. That is, in the future the REST api will allow `expand=prs,foos,bars`. At that time, what will `getCuration` look like? will it have prs, foos and bars as `params` and then a whole mess of tests to build up the REST query param list?

Instead, I suggest that `getCuration` take `params` as  { expand: ['prs'], state: 'pending' }. That elminates the plague of boolean params and makes the code API more similar to the REST API (and thus easier to doc and understand).

storrisi replied:
Ok, i see. Having the big picture of the future of the REST API gives to the implementation a different approach.
I will change this following your suggestions.

jeffmcaffer commented on src/api/clearlyDefined.js:
```diff
@@ -90,6 +91,12 @@ export function curate(token, spec) {
+ * Get details about a specific definition
+ * @param {*} token
+ * @param {*} entity
+ * @param {object} params can contain properties: expandPrs, if true include deeper information for each PR like the title, message
```
similar to above, there should be an `expand` arg that takes an array of enum values as to what to include.

jeffmcaffer commented:
Thanks @storrisi . I think we are on the same page here for the most part. I get your point about refactoring and doing things incrementally. The flip side is that when we create things that are not the way we want them, we create debt. Similarly, someone new to the codebase will look at that and think it is intended and mimic it.

How about you get the APIs in the right shape (no boolean args) and do a timeboxed amount of shaping up the component refactor (classes, proptypes, move up a level) and then we can merge and tweak in passes.

storrisi commented:
I've updated the PR enhancing the Refactoring of the Components, using a more flexible folder convention.
I've also updated the `expanded` and `state` properties.

jeffmcaffer commented:
Thanks @storrisi . Looking good.
